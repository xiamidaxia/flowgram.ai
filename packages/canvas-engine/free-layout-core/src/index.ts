/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

export * from './workflow-commands';
export * from './hooks';
export * from './utils';
export * from './typings';
export * from './entities';
export * from './constants';
export * from './entity-datas';
export * from './service';
export * from './workflow-document';
export * from './workflow-document-container-module';
export * from './workflow-lines-manager';
export * from './workflow-document-option';
