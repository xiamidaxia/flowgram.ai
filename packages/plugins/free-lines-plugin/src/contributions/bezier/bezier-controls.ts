/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { type IPoint } from '@flowgram.ai/utils';
import { LinePoint, LinePointLocation } from '@flowgram.ai/free-layout-core';

/**
 * Fork from: https://github.com/xyflow/xyflow/blob/main/packages/system/src/utils/edges/bezier-edge.ts
 * MIT License
 * Copyright (c) 2019-2024 webkid GmbH
 */
export function getBezierEdgeCenter(
  fromPos: IPoint,
  toPos: IPoint,
  fromControl: IPoint,
  toControl: IPoint
): IPoint {
  /*
   * cubic bezier t=0.5 mid point, not the actual mid point, but easy to calculate
   * https://stackoverflow.com/questions/67516101/how-to-find-distance-mid-point-of-bezier-curve
   */
  const x = fromPos.x * 0.125 + fromControl.x * 0.375 + toControl.x * 0.375 + toPos.x * 0.125;
  const y = fromPos.y * 0.125 + fromControl.y * 0.375 + toControl.y * 0.375 + toPos.y * 0.125;
  return {
    x,
    y,
  };
}

function getControlOffset(distance: number, curvature: number): number {
  if (distance >= 0) {
    return 0.5 * distance;
  }

  return curvature * 25 * Math.sqrt(-distance);
}

function getControlWithCurvature({
  location,
  x1,
  y1,
  x2,
  y2,
  curvature,
}: {
  location: LinePointLocation;
  curvature: number;
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}): IPoint {
  switch (location) {
    case 'left':
      return {
        x: x1 - getControlOffset(x1 - x2, curvature),
        y: y1,
      };
    case 'right':
      return {
        x: x1 + getControlOffset(x2 - x1, curvature),
        y: y1,
      };
    case 'top':
      return {
        x: x1,
        y: y1 - getControlOffset(y1 - y2, curvature),
      };
    case 'bottom':
      return {
        x: x1,
        y: y1 + getControlOffset(y2 - y1, curvature),
      };
  }
}

export function getBezierControlPoints(
  fromPos: LinePoint,
  toPos: LinePoint,
  curvature = 0.25
): { controls: [IPoint, IPoint]; center: IPoint } {
  const fromControl = getControlWithCurvature({
    location: fromPos.location,
    x1: fromPos.x,
    y1: fromPos.y,
    x2: toPos.x,
    y2: toPos.y,
    curvature,
  });
  const toControl = getControlWithCurvature({
    location: toPos.location,
    x1: toPos.x,
    y1: toPos.y,
    x2: fromPos.x,
    y2: fromPos.y,
    curvature,
  });
  const center = getBezierEdgeCenter(fromPos, toPos, fromControl, toControl);
  return {
    controls: [fromControl, toControl],
    center,
  };
}
