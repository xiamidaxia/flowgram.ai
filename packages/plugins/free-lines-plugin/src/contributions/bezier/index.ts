/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { Bezier } from 'bezier-js';
import { IPoint, Point, Rectangle } from '@flowgram.ai/utils';
import {
  WorkflowLineEntity,
  WorkflowLineRenderContribution,
  LinePoint,
  LineCenterPoint,
} from '@flowgram.ai/free-layout-core';
import { LineType } from '@flowgram.ai/free-layout-core';

import { posWithShrink, toRelative } from '../utils';
import { getBezierControlPoints } from './bezier-controls';

export interface BezierData {
  fromPos: IPoint;
  toPos: IPoint;
  bbox: Rectangle; // 外围矩形
  controls: IPoint[]; // 控制点
  bezier: Bezier;
  path: string;
  center: LineCenterPoint;
}

export class WorkflowBezierLineContribution implements WorkflowLineRenderContribution {
  public static type = LineType.BEZIER;

  public entity: WorkflowLineEntity;

  constructor(entity: WorkflowLineEntity) {
    this.entity = entity;
  }

  private data?: BezierData;

  public get path(): string {
    return this.data?.path ?? '';
  }

  public calcDistance(pos: IPoint): number {
    if (!this.data) {
      return Number.MAX_SAFE_INTEGER;
    }
    return Point.getDistance(pos, this.data.bezier.project(pos));
  }

  public get bounds(): Rectangle {
    if (!this.data) {
      return Rectangle.EMPTY;
    }
    return this.data.bbox;
  }

  get center() {
    return this.data?.center;
  }

  public update(params: { fromPos: LinePoint; toPos: LinePoint }): void {
    this.data = this.calcBezier(params.fromPos, params.toPos);
  }

  private calcBezier(fromPos: LinePoint, toPos: LinePoint): BezierData {
    const { controls, center } = getBezierControlPoints(
      fromPos,
      toPos,
      this.entity.uiState.curvature
    );
    const bezier = new Bezier([fromPos, ...controls, toPos]);
    const bbox = bezier.bbox();
    const bboxBounds = new Rectangle(
      bbox.x.min,
      bbox.y.min,
      bbox.x.max - bbox.x.min,
      bbox.y.max - bbox.y.min
    );
    const centerPoint = toRelative(center, bboxBounds);

    const path = this.getPath({ bbox: bboxBounds, fromPos, toPos, controls });

    this.data = {
      fromPos,
      toPos,
      bezier,
      bbox: bboxBounds,
      controls,
      path,
      center: {
        ...center,
        labelX: centerPoint.x,
        labelY: centerPoint.y,
      },
    };
    return this.data;
  }

  private getPath(params: {
    bbox: Rectangle;
    fromPos: LinePoint;
    toPos: LinePoint;
    controls: [IPoint, IPoint];
  }): string {
    const { bbox } = params;
    // 相对位置转换函数
    const fromPos = toRelative(params.fromPos, bbox);
    const toPos = toRelative(params.toPos, bbox);

    const controls = params.controls.map((c) => toRelative(c, bbox));
    const shrink = this.entity.uiState.shrink;

    const renderFromPos: IPoint = posWithShrink(fromPos, params.fromPos.location, shrink);

    const renderToPos: IPoint = posWithShrink(toPos, params.toPos.location, shrink);

    const controlPoints = controls.map((s) => `${s.x} ${s.y}`).join(',');
    return `M${renderFromPos.x} ${renderFromPos.y} C ${controlPoints}, ${renderToPos.x} ${renderToPos.y}`;
  }
}
