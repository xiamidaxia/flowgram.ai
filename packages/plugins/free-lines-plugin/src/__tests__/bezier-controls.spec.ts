/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { describe, it, expect } from 'vitest';
import { LinePoint } from '@flowgram.ai/free-layout-core';

import { getBezierControlPoints } from '../contributions/bezier/bezier-controls';

describe('bezier-controls', () => {
  it('getBezierControlPoints', () => {
    const fromPos: LinePoint = {
      x: 325,
      y: 41.5,
      location: 'bottom',
    };
    const toPos: LinePoint = {
      x: 225,
      y: 97,
      location: 'top',
    };
    expect(getBezierControlPoints(fromPos, toPos)).toMatchSnapshot();
  });
});
