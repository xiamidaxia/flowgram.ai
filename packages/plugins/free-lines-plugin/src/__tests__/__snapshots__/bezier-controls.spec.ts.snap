// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bezier-controls > getBezierControlPoints 1`] = `
{
  "center": {
    "x": 275,
    "y": 69.25,
  },
  "controls": [
    {
      "x": 325,
      "y": 69.25,
    },
    {
      "x": 225,
      "y": 69.25,
    },
  ],
}
`;
